File: Source/KSCrash/Recording/KSCrashReportStore.m
//
//  KSCrashReportStore.m
//
//  Created by Karl Stenerud on 2012-02-05.
//
//  Copyright (c) 2012 Karl Stenerud. All rights reserved.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall remain in place
// in this source code.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//


#import "KSCrashReportStore.h"

#import "ARCSafe_MemMgmt.h"
#import "KSCrashReportFields.h"
#import "KSJSONCodecObjC.h"
#import "KSSafeCollections.h"
#import "NSDictionary+Merge.h"
#import "NSError+SimpleConstructor.h"
#import "RFC3339DateTool.h"
#import "KSCrashDoctor.h"

//#define KSLogger_LocalLevel TRACE
#import "KSLogger.h"


// ============================================================================
#pragma mark - Constants -
// ============================================================================

#define kCrashReportSuffix @"-CrashReport-"
#define kRecrashReportSuffix @"-RecrashReport-"


// ============================================================================
#pragma mark - Meta Data -
// ============================================================================

/**
 * Metadata class to hold name and creation date for a file, with
 * default comparison based on the creation date (ascending).
 */
@interface KSCrashReportInfo: NSObject

@property(nonatomic,readonly,retain) NSString* reportID;
@property(nonatomic,readonly,retain) NSDate* creationDate;

+ (KSCrashReportInfo*) reportInfoWithID:(NSString*) reportID
                           creationDate:(NSDate*) creationDate;

- (id) initWithID:(NSString*) reportID creationDate:(NSDate*) creationDate;

- (NSComparisonResult) compare:(KSCrashReportInfo*) other;

@end

@implementation KSCrashReportInfo

@synthesize reportID = _reportID;
@synthesize creationDate = _creationDate;

+ (KSCrashReportInfo*) reportInfoWithID:(NSString*) reportID
                           creationDate:(NSDate*) creationDate
{
    return as_autorelease([[self alloc] initWithID:reportID
                                      creationDate:creationDate]);
}

- (id) initWithID:(NSString*) reportID creationDate:(NSDate*) creationDate
{
    if((self = [super init]))
    {
        _reportID = as_retain(reportID);
        _creationDate = as_retain(creationDate);
    }
    return self;
}

- (void) dealloc
{
    as_release(_reportID);
    as_release(_creationDate);
    as_superdealloc();
}

- (NSComparisonResult) compare:(KSCrashReportInfo*) other
{
    return [_creationDate compare:other->_creationDate];
}

@end


// ============================================================================
#pragma mark - Main Class -
// ============================================================================

@interface KSCrashReportStore ()

@property(nonatomic,readwrite,retain) NSString* path;
@property(nonatomic,readwrite,retain) NSString* bundleName;

@end


@implementation KSCrashReportStore

#pragma mark Properties

@synthesize path = _path;
@synthesize bundleName = _bundleName;


#pragma mark Construction

+ (KSCrashReportStore*) storeWithPath:(NSString*) path
{
    return as_autorelease([[self alloc] initWithPath:path]);
}

- (id) initWithPath:(NSString*) path
{
    if((self = [super init]))
    {
        self.path = path;
        self.bundleName = [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleName"];
    }
    return self;
}

- (void) dealloc
{
    as_release(_path);
    as_release(_bundleName);
    as_superdealloc();
}

#pragma mark API

- (NSArray*) reportIDs
{
    NSError* error = nil;
    NSFileManager* fm = [NSFileManager defaultManager];
    NSArray* filenames = [fm contentsOfDirectoryAtPath:self.path error:&error];
    if(filenames == nil)
    {
        KSLOG_ERROR(@"Could not get contents of directory %@: %@", self.path, error);
        return nil;
    }

    NSMutableArray* reports = [NSMutableArray arrayWithCapacity:[filenames count]];
    for(NSString* filename in filenames)
    {
        NSString* reportId = [self reportIDFromFilename:filename];
        if(reportId != nil)
        {
            NSString* fullPath = [self.path stringByAppendingPathComponent:filename];
            NSDictionary* fileAttribs = [fm attributesOfItemAtPath:fullPath error:&error];
            if(fileAttribs == nil)
            {
                KSLOG_ERROR(@"Could not read file attributes for %@: %@", fullPath, error);
            }
            else
            {
                [reports addObject:[KSCrashReportInfo reportInfoWithID:reportId
                                                          creationDate:[fileAttribs valueForKey:NSFileCreationDate]]];
            }
        }
    }
    [reports sortUsingSelector:@selector(compare:)];

    NSMutableArray* sortedIDs = [NSMutableArray arrayWithCapacity:[reports count]];
    for(KSCrashReportInfo* info in reports)
    {
        [sortedIDs addObject:info.reportID];
    }
    return sortedIDs;
}

- (NSUInteger) reportCount
{
    return [[self reportIDs] count];
}

- (NSDictionary*) reportWithID:(NSString*) reportID
{
    NSError* error = nil;
    NSMutableDictionary* crashReport = [self readReport:[self pathToCrashReportWithID:reportID] error:&error];
    if(error != nil)
    {
        KSLOG_ERROR(@"Encountered error loading crash report %@: %@", reportID, error);
    }
    if(crashReport == nil)
    {
        KSLOG_ERROR(@"Could not load crash report");
        return nil;
    }
    NSMutableDictionary* recrashReport = [self readReport:[self pathToRecrashReportWithID:reportID] error:nil];
    [crashReport setObjectIfNotNil:recrashReport forKey:@KSCrashField_RecrashReport];

    return crashReport;
}

- (NSArray*) allReports
{
    NSArray* reportIDs = [self reportIDs];
    NSMutableArray* reports = [NSMutableArray arrayWithCapacity:[reportIDs count]];
    for(NSString* reportID in reportIDs)
    {
        NSDictionary* report = [self reportWithID:reportID];
        if(report != nil)
        {
            [reports addObject:report];
        }
    }

    return reports;
}

- (void) deleteReportWithID:(NSString*) reportID
{
    NSError* error = nil;
    NSString* filename = [self pathToCrashReportWithID:reportID];

    [[NSFileManager defaultManager] removeItemAtPath:filename error:&error];
    if(error != nil)
    {
        KSLOG_ERROR(@"Could not delete file %@: %@", filename, error);
    }

    // Don't care if this succeeds or not since it may not exist.
    [[NSFileManager defaultManager] removeItemAtPath:[self pathToRecrashReportWithID:reportID]
                                               error:&error];
}

- (void) deleteAllReports
{
    for(NSString* reportID in [self reportIDs])
    {
        [self deleteReportWithID:reportID];
    }
}

- (void) pruneReportsLeaving:(int) numReports
{
    NSArray* reportIDs = [self reportIDs];
    int deleteCount = (int)[reportIDs count] - numReports;
    for(int i = 0; i < deleteCount; i++)
    {
        [self deleteReportWithID:[reportIDs objectAtIndex:(NSUInteger)i]];
    }
}


#pragma mark Utility

- (NSMutableDictionary*) fixupCrashReport:(NSDictionary*) report
{
    if(![report isKindOfClass:[NSDictionary class]])
    {
        KSLOG_ERROR(@"Report should be a dictionary, not %@", [report class]);
        return nil;
    }

    NSMutableDictionary* mutableReport = as_autorelease([report mutableCopy]);
    NSMutableDictionary* mutableInfo = as_autorelease([[report objectForKey:@KSCrashField_Report] mutableCopy]);
    [mutableReport setObjectIfNotNil:mutableInfo forKey:@KSCrashField_Report];

    // Timestamp gets stored as a unix timestamp. Convert it to rfc3339.
    [self convertTimestamp:@KSCrashField_Timestamp inReport:mutableInfo];

    [self mergeDictWithKey:@KSCrashField_SystemAtCrash
           intoDictWithKey:@KSCrashField_System
                  inReport:mutableReport];

    [self mergeDictWithKey:@KSCrashField_UserAtCrash
           intoDictWithKey:@KSCrashField_User
                  inReport:mutableReport];

    NSMutableDictionary* crashReport = as_autorelease([[report objectForKey:@KSCrashField_Crash] mutableCopy]);
    [mutableReport setObjectIfNotNil:crashReport forKey:@KSCrashField_Crash];
    KSCrashDoctor* doctor = [KSCrashDoctor doctor];
    [crashReport setObjectIfNotNil:[doctor diagnoseCrash:report] forKey:@KSCrashField_Diagnosis];

    return mutableReport;
}

- (void) mergeDictWithKey:(NSString*) srcKey
          intoDictWithKey:(NSString*) dstKey
                 inReport:(NSMutableDictionary*) report
{
    NSDictionary* srcDict = [report objectForKey:srcKey];
    if(srcDict == nil)
    {
        // It's OK if the source dict didn't exist.
        return;
    }
    if(![srcDict isKindOfClass:[NSDictionary class]])
    {
        KSLOG_ERROR(@"'%@' should be a dictionary, not %@", srcKey, [srcDict class]);
        return;
    }

    NSDictionary* dstDict = [report objectForKey:dstKey];
    if(dstDict == nil)
    {
        dstDict = [NSDictionary dictionary];
    }
    if(![dstDict isKindOfClass:[NSDictionary class]])
    {
        KSLOG_ERROR(@"'%@' should be a dictionary, not %@", dstKey, [dstDict class]);
        return;
    }

    [report setObjectIfNotNil:[srcDict mergedInto:dstDict] forKey:dstKey];
    [report removeObjectForKey:srcKey];
}

- (void) convertTimestamp:(NSString*) key
                 inReport:(NSMutableDictionary*) report
{
    NSNumber* timestamp = [report objectForKey:key];
    if(timestamp == nil)
    {
        KSLOG_ERROR(@"entry '%@' not found", key);
        return;
    }
    if(![timestamp isKindOfClass:[NSNumber class]])
    {
        KSLOG_ERROR(@"'%@' should be a number, not %@", key, [key class]);
        return;
    }
    [report setValue:[RFC3339DateTool stringFromUNIXTimestamp:[timestamp unsignedLongLongValue]]
              forKey:key];
}

- (NSString*) crashReportFilenameWithID:(NSString*) reportID
{
    return [NSString stringWithFormat:@"%@" kCrashReportSuffix "%@.json",
            self.bundleName, reportID];
}

- (NSString*) recrashReportFilenameWithID:(NSString*) reportID
{
    return [NSString stringWithFormat:@"%@" kRecrashReportSuffix "%@.json",
            self.bundleName, reportID];
}

- (NSString*) reportIDFromFilename:(NSString*) filename
{
    if([filename length] == 0)
    {
        return nil;
    }

    NSString* prefix = [NSString stringWithFormat:@"%@" kCrashReportSuffix,
                        self.bundleName];
    NSString* suffix = @".json";
    if([filename rangeOfString:prefix].location == 0 &&
       [filename rangeOfString:suffix].location != NSNotFound)
    {
        NSUInteger prefixLength = [prefix length];
        NSUInteger suffixLength = [suffix length];
        NSRange range = NSMakeRange(prefixLength, [filename length] - prefixLength - suffixLength);
        return [filename substringWithRange:range];
    }
    return nil;
}

- (NSString*) pathToCrashReportWithID:(NSString*) reportID
{
    NSString* filename = [self crashReportFilenameWithID:reportID];
    return [self.path stringByAppendingPathComponent:filename];
}

- (NSString*) pathToRecrashReportWithID:(NSString*) reportID
{
    NSString* filename = [self recrashReportFilenameWithID:reportID];
    return [self.path stringByAppendingPathComponent:filename];
}

- (NSMutableDictionary*) readReport:(NSString*) path error:(NSError* __autoreleasing *) error
{
    if(path == nil)
    {
        [NSError fillError:error withDomain:[[self class] description]
                      code:0
               description:@"Path is nil"];
        return nil;
    }

    NSData* jsonData = [NSData dataWithContentsOfFile:path options:0 error:error];
    if(jsonData == nil)
    {
        return nil;
    }

    NSMutableDictionary* report = [self fixupCrashReport:[KSJSONCodec decode:jsonData
                                                                     options:KSJSONDecodeOptionIgnoreNullInArray |
                                                          KSJSONDecodeOptionIgnoreNullInObject |
                                                          KSJSONDecodeOptionKeepPartialObject
                                                                       error:error]];
    if(error != nil && *error != nil)
    {
        
        KSLOG_ERROR(@"Error decoding JSON data from %@: %@", path, *error);
        [report setObject:[NSNumber numberWithBool:YES] forKey:@KSCrashField_Incomplete];
    }

    return report;
}

@end

File: Source/KSCrash/Recording/KSCrashReportStore.c
//
//  KSCrashReportStore.c
//
//  Created by Karl Stenerud on 2012-02-05.
//
//  Copyright (c) 2012 Karl Stenerud. All rights reserved.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall remain in place
// in this source code.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//

#include "KSCrashReportStore.h"
#include "KSLogger.h"

#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>


static const int g_maxReports = 5;
static int64_t g_nextCrashID;
static int64_t g_nextUserReportID;
static char g_appName[KSCRS_MAX_PATH_LENGTH];
static char g_reportsPath[KSCRS_MAX_PATH_LENGTH];
static pthread_mutex_t g_mutex = PTHREAD_MUTEX_INITIALIZER;

static int compareInt64(const void* a, const void* b)
{
    int64_t diff = *(int64_t*)a - *(int64_t*)b;
    if(diff < 0)
    {
        return -1;
    }
    else if(diff > 0)
    {
        return 1;
    }
    return 0;
}

static bool makePath(const char* absolutePath)
{
    bool isSuccessful = false;
    char* pathCopy = strdup(absolutePath);
    for(char* ptr = pathCopy+1; *ptr != '\0';ptr++)
    {
        if(*ptr == '/')
        {
            *ptr = '\0';
            if(mkdir(pathCopy, S_IRWXU) < 0 && errno != EEXIST)
            {
                KSLOG_ERROR("Could not create directory %s: %s", pathCopy, strerror(errno));
                goto done;
            }
            *ptr = '/';
        }
    }
    if(mkdir(pathCopy, S_IRWXU) < 0 && errno != EEXIST)
    {
        KSLOG_ERROR("Could not create directory %s: %s", pathCopy, strerror(errno));
        goto done;
    }
    isSuccessful = true;
    
done:
    free(pathCopy);
    return isSuccessful;
}

static void removeFile(const char* path, bool mustExist)
{
    if(remove(path) < 0)
    {
        if(mustExist || errno != ENOENT)
        {
            KSLOG_ERROR("Could not delete %s: %s", path, strerror(errno));
        }
    }
}

static int readFile(const char* path, char** bufferPtr)
{
    int length = 0;
    char* buffer = NULL;
    *bufferPtr = NULL;
    
    const int fd = open(path, O_RDONLY);
    if(fd < 0)
    {
        if(errno != ENOENT)
        {
            KSLOG_ERROR("Could not open file %s: %s", path, strerror(errno));
        }
        goto done;
    }
    
    struct stat sb = {0};
    if(fstat(fd, &sb) < 0)
    {
        KSLOG_ERROR("Could not stat file %s: %s", path, strerror(errno));
        goto done;
    }
    
    length = (int)sb.st_size;
    if(length == 0)
    {
        KSLOG_ERROR("File %s is empty", path);
        goto done;
    }
    
    buffer = malloc((size_t)length);
    if(buffer == NULL)
    {
        KSLOG_ERROR("Could allocate %d bytes for file %s: %s", length, path, strerror(errno));
        goto done;
    }
    
    int bytesRead = read(fd, buffer, (size_t)length);
    if(bytesRead < 0)
    {
        KSLOG_ERROR("Error reading from file %s: %s", path, strerror(errno));
        goto done;
    }
    if(bytesRead != length)
    {
        KSLOG_ERROR("Expected to read %d bytes from file %s, but only got %ll", path, length, bytesRead);
        length = bytesRead;
    }
    *bufferPtr = buffer;
    buffer = NULL;
    
done:
    if(fd >= 0)
    {
        close(fd);
    }
    if(buffer != NULL)
    {
        free(buffer);
    }
    return length;
}

static void getCrashReportPathByID(int64_t id, char* pathBuffer)
{
    snprintf(pathBuffer, KSCRS_MAX_PATH_LENGTH, "%s/%s-report-%016llx.json", g_reportsPath, g_appName, id);
    
}

static int64_t getReportIDFromFilename(const char* filename)
{
    char scanFormat[100];
    sprintf(scanFormat, "%s-report-%%llx.json", g_appName);
    
    int64_t reportID = 0;
    sscanf(filename, scanFormat, &reportID);
    return reportID;
}

static void deleteReportWithID(int64_t id)
{
    char path[KSCRS_MAX_PATH_LENGTH];
    getCrashReportPathByID(id, path);
    removeFile(path, true);
}

static int getReportCount()
{
    int count = 0;
    DIR* dir = opendir(g_reportsPath);
    if(dir == NULL)
    {
        KSLOG_ERROR("Could not open directory %s", g_reportsPath);
        goto done;
    }
    struct dirent* ent;
    while((ent = readdir(dir)) != NULL)
    {
        if(getReportIDFromFilename(ent->d_name) > 0)
        {
            count++;
        }
    }

done:
    if(dir != NULL)
    {
        closedir(dir);
    }
    return count;
}

static int getReportIDs(int64_t* reportIDs, int count)
{
    int index = 0;
    DIR* dir = opendir(g_reportsPath);
    if(dir == NULL)
    {
        KSLOG_ERROR("Could not open directory %s", g_reportsPath);
        goto done;
    }

    struct dirent* ent;
    while((ent = readdir(dir)) != NULL && index < count)
    {
        int64_t reportID = getReportIDFromFilename(ent->d_name);
        if(reportID > 0)
        {
            reportIDs[index++] = reportID;
        }
    }

    qsort(reportIDs, (size_t)count, sizeof(reportIDs[0]), compareInt64);

done:
    if(dir != NULL)
    {
        closedir(dir);
    }
    return index;
}

static void pruneReports()
{
    int reportCount = getReportCount();
    if(reportCount > g_maxReports)
    {
        int64_t reportIDs[reportCount];
        reportCount = getReportIDs(reportIDs, reportCount);
        
        for(int i = 0; i < reportCount - g_maxReports; i++)
        {
            deleteReportWithID(reportIDs[i]);
        }
    }
}

static void initializeIDs()
{
    time_t rawTime;
    time(&rawTime);
    struct tm time;
    gmtime_r(&rawTime, &time);
    int64_t baseID = (int64_t)time.tm_sec
                     + (int64_t)time.tm_min * 61
                     + (int64_t)time.tm_hour * 61 * 60
                     + (int64_t)time.tm_yday * 61 * 60 * 24
                     + (int64_t)time.tm_year * 61 * 60 * 24 * 366;
    baseID <<= 23;

    g_nextCrashID = baseID | 0x400000;
    g_nextUserReportID = baseID;
}


// Public API

void kscrs_initialize(const char* appName, const char* reportsPath)
{
    pthread_mutex_lock(&g_mutex);
    strncpy(g_appName, appName, sizeof(g_appName));
    strncpy(g_reportsPath, reportsPath, sizeof(g_reportsPath));
    makePath(reportsPath);
    pruneReports();
    initializeIDs();
    pthread_mutex_unlock(&g_mutex);
}

void kscrs_getCrashReportPath(char* crashReportPathBuffer)
{
    pthread_mutex_lock(&g_mutex);
    getCrashReportPathByID(g_nextCrashID, crashReportPathBuffer);
    pthread_mutex_unlock(&g_mutex);
}

int kscrs_getReportCount()
{
    pthread_mutex_lock(&g_mutex);
    int count = getReportCount();
    pthread_mutex_unlock(&g_mutex);
    return count;
}

int kscrs_getReportIDs(int64_t* reportIDs, int count)
{
    pthread_mutex_lock(&g_mutex);
    count = getReportIDs(reportIDs, count);
    pthread_mutex_unlock(&g_mutex);
    return count;
}

void kscrs_readReport(int64_t reportID, char** reportPtr, int* reportLengthPtr)
{
    pthread_mutex_lock(&g_mutex);
    char path[KSCRS_MAX_PATH_LENGTH];
    getCrashReportPathByID(reportID, path);
    *reportLengthPtr = readFile(path, reportPtr);
    pthread_mutex_unlock(&g_mutex);
}

void kscrs_addUserReport(const char* report, int reportLength)
{
    pthread_mutex_lock(&g_mutex);
    char crashReportPath[KSCRS_MAX_PATH_LENGTH];
    getCrashReportPathByID(g_nextUserReportID, crashReportPath);
    g_nextUserReportID++;

    int fd = open(crashReportPath, O_WRONLY | O_CREAT, 0644);
    if(fd < 0)
    {
        KSLOG_ERROR("Could not open file %s: %s", crashReportPath, strerror(errno));
        goto done;
    }

    int bytesWritten = write(fd, report, (size_t)reportLength);
    if(bytesWritten < 0)
    {
        KSLOG_ERROR("Could not write to file %s: %s", crashReportPath, strerror(errno));
        goto done;
    }
    else if(bytesWritten < (ssize_t)reportLength)
    {
        KSLOG_ERROR("Expected to write %lull bytes to file %s, but only wrote %ll", crashReportPath, reportLength, bytesWritten);
    }

done:
    if(fd >= 0)
    {
        close(fd);
    }
    pthread_mutex_unlock(&g_mutex);
}

void kscrs_deleteAllReports()
{
    pthread_mutex_lock(&g_mutex);
    DIR* dir = opendir(g_reportsPath);
    if(dir != NULL)
    {
        char pathBuffer[KSCRS_MAX_PATH_LENGTH];
        snprintf(pathBuffer, sizeof(pathBuffer), "%s/", g_reportsPath);
        char* pathPtr = pathBuffer + strlen(pathBuffer);
        int pathRemainingLength = (int)sizeof(pathBuffer) - (int)(pathPtr - pathBuffer);
        
        for(bool mustRescan = true; mustRescan; mustRescan = false)
        {
            rewinddir(dir);
            struct dirent* ent;
            while((ent = readdir(dir)))
            {
                strncpy(pathPtr, ent->d_name, pathRemainingLength);
                removeFile(pathBuffer, false);
                mustRescan = true;
            }
        }
        closedir (dir);
    }
    else
    {
        KSLOG_ERROR("Could not open directory %s", g_reportsPath);
    }
    pthread_mutex_unlock(&g_mutex);
}


// Internal API

void kscrsi_incrementCrashReportIndex()
{
    g_nextCrashID++;
}

int64_t kscrsi_getNextCrashReportID()
{
    return g_nextCrashID;
}

int64_t kscrsi_getNextUserReportID()
{
    return g_nextUserReportID;
}
